openjpa-kernel/src/main/java/org/apache/openjpa/kernel/BrokerImpl.java

```diff
@@ -2149,8 +2149,16 @@ protected void flush(int reason) {
             // dependents
             _flags |= FLAG_DEREFDELETING;
             if (flush && _derefCache != null && !_derefCache.isEmpty()) {
-                for (Iterator<StateManagerImpl> itr = _derefCache.iterator(); itr.hasNext();)
-                    deleteDeref(itr.next());
+                Set<StateManagerImpl> statesMarkedForDelete = new HashSet<>();
+                // mark for delete all elements in deref, otherwise in some situations it
+                // throws ConcurrentModificationException
+                statesMarkedForDelete.addAll(_derefCache);
+                if(statesMarkedForDelete.size()>0) {
```

Comment:
This if can be safely removed: the subsequent foreach will skip in case of empty set

Reply:
Correct, thanks. removed